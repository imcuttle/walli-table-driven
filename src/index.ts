/**
 * The table driven engine which respect walli
 * @author imcuttle
 */
import { Verifiable, leq } from 'walli'

export type WhereRule<T = any> =
  | {
      [P in keyof T]?: T[P] | Verifiable
    }
  | Verifiable

export type TableRuleResult<T, R> = R | ((source: T, check: (rule: WhereRule<T>) => boolean) => R)
export interface TableRuleOptions {
  shouldContinue?: boolean
}

export interface StrictTableRule<T = any, R = any> {
  where: WhereRule<T>
  result?: TableRuleResult<T, R>
  options?: TableRuleOptions
}

export type TableRule<T = any, R = any> =
  | StrictTableRule<T, R>
  | [WhereRule<T>, TableRuleResult<T, R>, TableRuleOptions?]

export function toStrictTableRule(rule): StrictTableRule {
  if (Array.isArray(rule)) {
    return {
      where: rule[0],
      result: rule[1],
      options: rule[2]
    }
  }
  return rule
}

export function toVerifiable(rule: any = {}) {
  let verifiable: Verifiable
  if (rule instanceof Verifiable) {
    verifiable = rule
  } else {
    verifiable = leq(rule)
  }
  return verifiable
}

export function walliTableDrivenQuery<T = any, R = any>(
  value: T,
  tableRule: StrictTableRule<T, R> | TableRule<T, R>[]
):
  | {
      matched: false
      result: null
    }
  | {
      result: R
      rule: StrictTableRule<T, R> | StrictTableRule<T, R>[]
      matched: true
    } {
  if (Array.isArray(tableRule)) {
    const matchedList: any[] = []
    tableRule.some((rule) => {
      const strictRule = toStrictTableRule(rule)
      const { shouldContinue = false } = strictRule.options || {}
      const currentMatched = walliTableDrivenQuery(value, strictRule)
      if (currentMatched.matched) {
        matchedList.push(currentMatched)
        // 需要 merge，并且已经匹配过
        if (shouldContinue) {
          return false
        }
      }
      return currentMatched.matched
    })

    // tslint:disable-next-line:no-shadowed-variable
    const mergedResult = matchedList.reduce((acc, matched: any) => {
      if (!acc) {
        return matched.result
      }
      return {
        ...acc,
        ...matched.result
      }
    }, null)

    const res = {
      matched: !!matchedList.length,
      result: mergedResult as R
    }
    if (res.matched) {
      // @ts-ignore
      res.rule = matchedList
    }
    return res as any
  }

  const strictTableRule = toStrictTableRule(tableRule)
  const verifiable = toVerifiable(strictTableRule.where)

  if (verifiable.ok(value)) {
    let result = strictTableRule.result
    if (typeof strictTableRule.result === 'function') {
      const check = (where: WhereRule) => {
        return toVerifiable(where).ok(value)
      }
      result = strictTableRule.result(value, check)
    }
    return { result, rule: strictTableRule, matched: true }
  }
  return { result: null, matched: false }
}
