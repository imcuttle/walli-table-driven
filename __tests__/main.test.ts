/**
 * @file main
 * @author imcuttle
 * @date 2018/4/4
 */
import { walliTableDrivenQuery } from '../src'

describe('walliTableDriven', function () {
  it('should spec', function () {
    expect(
      walliTableDrivenQuery(
        {
          age: 20
        },
        [[{ age: 2 }, null]]
      )
    ).toMatchSnapshot()

    expect(
      walliTableDrivenQuery(
        {
          age: 20
        },
        [
          [{ age: 2 }, null],
          [{ age: 20 }, null]
        ]
      )
    ).toMatchSnapshot()

    expect(
      walliTableDrivenQuery(
        {
          age: 20
        },
        {
          where: { age: 20 }
        }
      )
    ).toMatchSnapshot()
  })
})
