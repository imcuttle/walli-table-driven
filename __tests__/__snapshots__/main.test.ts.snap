// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`walliTableDriven should spec 1`] = `
Object {
  "matched": false,
  "result": null,
}
`;

exports[`walliTableDriven should spec 2`] = `
Object {
  "matched": true,
  "result": null,
  "rule": Array [
    Object {
      "matched": true,
      "result": null,
      "rule": Object {
        "options": undefined,
        "result": null,
        "where": Object {
          "age": 20,
        },
      },
    },
  ],
}
`;
